lib/uplink/main.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink
+
+import (
+	"context"
+
+	"storj.io/storj/pkg/transport"
+
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/metainfo/kvmetainfo"
+	"storj.io/storj/pkg/miniogw"
+	"storj.io/storj/pkg/peertls/tlsopts"
+	"storj.io/storj/pkg/storj"
+	ul "storj.io/storj/uplink"
+)
+
+// Config holds the configs for the Uplink
+type Config struct {
+	// MaxBufferMem controls upload performance and is system-specific
+	MaxBufferMem int
+
+	// These should only be relevant for new files; these values for existing
+	// files should come from the metainfo index. It's unlikely these will ever
+	// change much.
+	EncBlockSize  int
+	MaxInlineSize int
+	SegmentSize   int64
+	TLSConfig     tlsopts.Config
+}
+
+// Session represents a specific access session.
+type Session struct {
+	TransportClient transport.Client
+	Gateway         *miniogw.Gateway
+}
+
+// Access is all of the access information an application needs to store and
+// retrieve data. Someone with a share may have no restrictions within a project
+// (can create buckets, list buckets, list files, upload files, delete files,
+// etc), may be restricted to a single bucket, may be restricted to a prefix
+// within a bucket, or may even be restricted to a single file within a bucket.
+// NB(dylan): You need an Access to start a Session
+type Access struct {
+	Permissions Macaroon
+
+	// TODO: these should be per-bucket somehow maybe? oh man what a nightmare
+	// Could be done via []Bucket with struct that has each of these
+	// PathCipher       storj.Cipher // i.e. storj.AESGCM
+	// EncPathPrefix    storj.Path
+	// Key              storj.Key
+	// EncryptionScheme storj.EncryptionScheme
+
+	// Something like this?
+	// TODO(dylan): Shouldn't actually use string, this is just a placeholder
+	// until a more precise type is figured out - probably type Bucket
+	Buckets map[string]BucketOpts
```

Comment:
This map should probably be indexed by a pair of (ProjectID, BucketName). Or I suppose we could say "you have to instantiate a new Uplink object for every different ProjectID you're working with" and then this could safely be just `string` for bucket names.

Reply:
That's not a bad idea, actually. I'll look into that. Making another TODO 👍